wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -0,0 +1,193 @@
+import           Cardano.Wallet.Kernel.Decrypt (decryptAddress)
+import           Cardano.Wallet.Kernel.Internal (WalletRestorationInfo (..),
+                     walletMeta, walletNode, walletRestorationTask,
+                     walletSubmission, wallets, wriCurrentSlot, wriTargetSlot)
+import           Cardano.Wallet.Kernel.NodeStateAdaptor (LockContext (..),
+                     SecurityParameter (..), getSecurityParameter,
+                     getSlotCount, withNodeState)
+import           Cardano.Wallet.Kernel.PrefilterTx (PrefilteredBlock, WalletKey,
+                     toHdAddressId)
+import qualified Cardano.Wallet.Kernel.Submission as Submission
+import           Cardano.Wallet.Kernel.Types (WalletId)
+
+import           Pos.Chain.Block (BlockHeader, Blund, HeaderHash, blockHeader,
+                     headerHash, mainBlockSlot)
+import           Pos.Chain.Txp (Utxo, getTotalCoinsInUtxo)
+import           Pos.Core (Address, Coin, SlotId, flattenSlotIdExplicit)
+import           Pos.Core.Txp (toaOut, txOutAddress)
+import           Pos.DB (MonadDBRead)
+import           Pos.DB.Block (getBlund, getFirstGenesisBlockHash,
+                     resolveForwardLink)
+import           Pos.DB.Class (getBlock)
+import           Pos.DB.Txp.Utxo (utxoSource)
+import           Pos.Wallet.Web.State (WAddressMeta (..))
+
+
+-- | Scan the node's current UTXO set for any that belong to this wallet. Use them
+--   to update the current checkpoint's UTXO set, and return the total 'Coin' value
+--   of the UTXO belonging to this wallet. At the same time, kick off a background
+--   thread that will asynchronously restore the wallet history.
+restoreWallet :: Kernel.PassiveWallet
+              -> HD.HdRoot
+              -> WalletKey
+              -> (Blund -> IO (Map HD.HdAccountId PrefilteredBlock, [TxMeta]))
+              -> IO Coin
+restoreWallet wallet hdRoot (wId, wdc) prefilter = do
+    -- Find all of the current UTXO that this wallet owns.
+    mine <- withNodeState (wallet ^. walletNode) (\_lock -> filterMyUtxo)
+
+    -- Create the wallet and populate the accounts we've seen.
+    let db = wallet ^. wallets
+        utxoByAcct = map (\(u, a, wam) ->
+                                 let hdAddr = initHdAddress (toHdAddressId wId wam) (InDb a)
+                                     hdAddrId = hdAddr ^. HD.hdAddressId
+                                 in (hdAddr ^. HD.hdAddressAccountId, (u, [(hdAddrId, a)]))) mine
+
+    csr <- getChainStateRestoration (wallet ^. walletNode) NotYetLocked
+    whenJust (csrCurrent csr) $ \brief -> do
+        -- Create the wallet
+        let tgtTip  = cbTip brief
+            tgtSlot = cbSlotId brief
+        void $ update db (RestoreHdWallet hdRoot tgtSlot (M.fromListWith (<>) utxoByAcct))
+
+        -- Begin restoring the wallet history in the background.
+        restoreTask <- async (restoreWalletHistoryAsync wallet wId tgtTip tgtSlot prefilter)
+        link restoreTask
```

Comment:
Is the concern about the `Async` running but due to an exception on this thread the `Async` reference becomes orphaned? I couldn't figure out how to avoid this, unfortunately..